src/pairingheap.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Datastructures: GAP package providing common datastructures.
+ * Licensed under the GPL 2 or later.
+ *
+ * Implementation of a binary heap.
+ *
+ * Binary heaps are of course pretty standard data structures.
+ * However, a few design choices are possible. The implementation
+ * below has been influenced by this StackOverflow answer:
+ *   <https://stackoverflow.com/questions/6531543>
+ */
+
+#include "src/compiled.h"
+#include "pairingheap.h"
+
+#define DS_BINARYHEAP_ISLESS(heap) ELM_PLIST(heap, 1)
+#define DS_BINARYHEAP_DATA(heap) ELM_PLIST(heap, 2)
+
+
+static Obj SumSmallInt(Obj a, Obj b)
+{
+    GAP_ASSERT(IS_INTOBJ(a));
+    GAP_ASSERT(IS_INTOBJ(b));
+    Obj sum;
+    if (!SUM_INTOBJS(sum, a, b))
+        ErrorMayQuit("PANIC: counter overflow in pairing heap code", 0, 0);
+    return sum;
+}
+
+Obj DS_merge_pairs(Obj self, Obj isLess, Obj heaps)
+{
+    if (!IS_DENSE_PLIST(heaps))
+        ErrorQuit("<heaps> is not a dense plist", 0L, 0L);
+
+    const UInt len = LEN_PLIST(heaps);
+    Obj res;
+
+    if (len == 0) {
+        res = NEW_PLIST(T_PLIST_CYC_SSORT, 3);
+        SET_LEN_PLIST(res, 3);
+        SET_ELM_PLIST(res, 1, INTOBJ_INT(0));
+        SET_ELM_PLIST(res, 2, INTOBJ_INT(0));
+        SET_ELM_PLIST(res, 3, INTOBJ_INT(0));
+        return res;
+    }
+    
+    if (len == 1) {
+        return ELM_PLIST(heaps, 1);
+    }
+    
+    const Int useLt = (isLess == LtOper);
+    res = heaps;
+    UInt k = len;
+    UInt s = 1;
+    UInt i;
+    while (k > 1) {
+        const UInt r = k & 1;
+        const UInt old_s = s;
+        k >>= 1;
+        s <<= 1;
+        for (i = s; i <= k*s; i += s) {
+            GAP_ASSERT(i <= LEN_PLIST(res));
+            Obj x = ELM_PLIST(res, i - old_s);
+            Obj y = ELM_PLIST(res, i);
+            GAP_ASSERT(IS_DENSE_PLIST(x));
+            GAP_ASSERT(IS_DENSE_PLIST(y));
+
+            const Obj sum = SumSmallInt(ELM_PLIST(x, 2), ELM_PLIST(y, 2));
```

Comment:
Good idea. We don't do that on the GAP level, for some extra speed, but giving names on the C level is of course free.